### PR TITLE
POS-58 Implement liveness issue test proof

### PR DESCRIPTION
**Pull request by Stefan-Ethernal**
PR is follow up of stale PR https://github.com/0xPolygon/pbft-consensus/pull/19

This PR implements integration test that proves that PolyBFT consensus algorithm (inspired by IBFT) has liveness issues. Test is currently deliberately failing, because it expects that there are no errors, however timeout error is triggered by test framework since no consensus is reached after predefined time.

Test is implemented by having a scenario, discussed in [Correctness Analysis of Istanbul Byzantine Fault Tolerance paper, Chapter 7.1, Case 1](https://arxiv.org/pdf/1901.07160.pdf) on mind:
Note: for simplicity, imagine that we have validator set which has 5 validators {a1, a2, a3, a4 and a5}. Validator nodes are not named exactly this way within the test, but the point is there.

- a3 is proposer and proposes block B (sends PREPARE and PRE-PREPARE messages),
- a1 and a2 receive PRE-PREPARE from all the validator nodes and send PREPARE messages. a3, a4 and a5 don't receive either message (e.g. transient network issues),
- a1 and a2 locked on proposal B, since they have received enough PREPARE messages (2*F+1, where F is number of max faulty nodes),
- since there are only PREPARE messages from a1 and a2 and no COMMIT messages, when timeout elapses, all the nodes move to ROUND CHANGE state,
- when they eventually gossip enough ROUND CHANGE messages, they agree upon moving to the next round where a new proposer (e.g. a4) gets calculated,
- a4 proposes new proposal B' which is different from the one a1 and a2 got locked on,
- a3, a4 and a5 all respond with PREPARE message and lock on B' proposal (whereas a1 and a2 respond with ROUND CHANGE, since they are already locked on proposal B),
- before responding with COMMIT messages, some of those nodes becomes faulty (e.g. a3) and only a4 and a5 are able to send COMMIT messages. So we don't have enough COMMIT messages in order to move to COMMIT state (and eventually make a final proposal).

We have situation where a1 and a2 make partition 1 (locked on proposal B) and a3 and a5 (locked on proposal B'). The faulty a4 node is lacking in order to have quorum.

### CODE REVIEW

File: e2e/framework.go [hunk 1]
```diff
@@ -422,17 +425,16 @@ func (f *fsm) IsStuck(num uint64) (uint64, bool) {
 }
 
 func (f *fsm) BuildProposal() (*pbft.Proposal, error) {
+	// make different proposal for each sequence/round
+	prop := make([]byte, ProposalLength)
```
Comment (ferranbt) [inline e2e/framework.go:429]: Please revert this change. This is just a test, we do not need to do this here.

File: e2e/transport.go [hunk 2]
```diff
@@ -42,10 +42,41 @@ func (t *transport) Gossip(msg *pbft.MessageReq) error {
 }
 
 type transportHook interface {
-	Connects(from, to pbft.NodeID) bool
```
Comment (ferranbt) [inline e2e/transport.go:45]: Please revert this change. I am fine changing the naming of the interface but not as part of a PR that is addressing something else.

File: e2e/e2e_partition_test.go [hunk 3]
```diff
@@ -1,14 +1,133 @@
...
+	hook := newGenericGossipTransport(livenessGossipHandler)
+
+	c := newPBFTCluster(t, "liveness_issue", "A", nodesCnt, hook)
+	c.Start()
+	defer c.Stop()
+
+	err := c.WaitForHeight(3, 5*time.Minute)
+
+	// Query nodes and make appropriate assertions
+	nodeSubsetA := []pbft.NodeID{"A_0", "A_1", "A_2"}
```
Comment (ferranbt) [inline e2e/e2e_partition_test.go:81]: Please revert this change as well. We do not need to do an extra check to see if there are two locked proposals, it is enough for me if we log it. This is the idea behind the integration tests, to test some specific multi node situation and wait for the chain to finish. I do not want to over complicate it this way.

File: e2e/e2e_partition_test.go [hunk 4]
```diff
@@ -1,14 +1,133 @@
...
+	nodeA0 := c.nodes[string(nodeSubsetA[0])]
+	nodeA3 := c.nodes[string(nodeSubsetB[0])]
+	queryNodesFromSubset(t, c.nodes, faultyNodeId, nodeA0, nodeSubsetA)
+	queryNodesFromSubset(t, c.nodes, faultyNodeId, nodeA3, nodeSubsetB)
+
+	assert.NoError(t, err)
+}
+
+// Helper function determining whether a message should be gossiped, based on round message flow.
+func shouldGossipBasedOnMsgFlowMap(flowMap map[uint64]roundMetadata, msg *pbft.MessageReq, senderId pbft.NodeID, receiverId pbft.NodeID) bool {
```
Comment (ferranbt) [inline e2e/e2e_partition_test.go:92]: Could you add this function as a method of `roundMetadata`?
```
func (r *roundMetadata) shouldGossipBasedOnMsgFlowMap...
```

**Review by ferranbt — CHANGES_REQUESTED** [review]

File: e2e/e2e_partition_test.go (hunk 4 above)
Comment (Stefan-Ethernal) [reply e2e/e2e_partition_test.go]: Seems like it more suits to the `genericGossipTransport`, so I have moved it there. Hope that you don't mind.

**Review by ferranbt — APPROVED** [review]
LGTM! Lets merge this and work on the fix.
